docs/design/agentcube-proposal.md

```diff
@@ -0,0 +1,475 @@
+- Simple developer-facing abstraction  
+  Expose a clean API/CRD or SDK so application teams can request “an agent/runtime with X capabilities” without dealing with pods, nodes, or cluster internals.
+- Efficient use of shared infra (CPU/GPU)  
+  Integrate with platforms like Volcano and DRA to schedule agent workloads efficiently on heterogeneous resources (CPUs, GPUs, accelerators), minimizing waste.
+- Observability and operations for platform teams  
+  Provide metrics, logs, and controls so platform/SRE teams can monitor, tune, and govern large-scale agent and interpreter usage.
+
+### **Non-Goals**
+
+- Not an LLM or agent framework  
+  AgentCube does not aim to replace LLMs, agent libraries, or orchestration frameworks (e.g., LangChain, LlamaIndex). It focuses on runtime and infrastructure, not model logic.
+- Not a replacement for all Kubernetes workloads  
+  It is not intended to replace standard Deployments/Jobs for long-running services or classic batch jobs. It targets ephemeral, high-churn agent and code-execution workloads.
+- Not a multi-cloud control plane (yet)  
+  The initial focus is on running efficiently within Kubernetes clusters. Cross-cloud control and federation are out of scope for the first phase.
+
+## Use Cases
+
+1.  Interactive AI agents with low-latency startup
+    - Chat- and tool-based agents that must start quickly in response to user requests.
+    - On-demand provisioning of agent runtimes, with aggressive cold-start optimization and reuse where possible.
+2.  Secure, isolated code execution / code interpreter sessions
+    - Notebooks, “Run this code” buttons, and REPL-like experiences where user code should be executed in short-lived, isolated sandboxes.
+    - Per-session environments with strict resource limits, policy controls, and clear lifecycle management on Kubernetes.
+3.  Serverless-style agent orchestration
+    - Treating each agent invocation as a lightweight, event-driven workload, scheduled on demand and scaled to zero when idle.
+    - Integrating with existing Kubernetes primitives (autoscaling, scheduling, observability) while hiding complexity from application developers.
+4.  Multi-tenant AI platforms
+    - Platform teams providing standardized agent and interpreter runtimes to multiple product teams.
+    - Central management of runtime images, dependencies, security policies, and quota, while allowing tenants to customize tools and workflows.
+
+## Overall Architecture
+
+![](./images/agentcube.svg)
+
+### Concepts
+
+1.  **API: AgentRuntime**
+
+The CRD specifies the template to build agent runtime sandboxes and also can be used to specify the port, protocol, and url path that agent will be running with.
+
+2.  **API: CodeInterpreter**
+
+The CRD specifies the template to build code interpreter sandboxes and also can be used to specify the port, protocol, and url path that code interpreter will be running with.
```

Comment:
It would be better to explain why we need `CodeInterpreter` that we already have `AgentRuntime`.